src/main.rs

```diff
@@ -56,24 +64,47 @@ struct StepFailed;
 #[fail(display = "Cannot find the user base directories")]
 struct NoBaseDirectories;
 
-fn execute<'a, F, M>(func: F, terminal: &mut Terminal) -> Option<(M, bool)>
+#[derive(Fail, Debug)]
+#[fail(display = "Process Interrupted")]
+struct Interrupted;
+
+struct ExecutionContext {
+    terminal: Terminal,
+}
+
+fn execute<'a, F, M>(func: F, execution_context: &mut ExecutionContext) -> Result<Option<(M, bool)>, Interrupted>
 where
     M: Into<Cow<'a, str>>,
     F: Fn(&mut Terminal) -> Option<(M, bool)>,
 {
-    while let Some((key, success)) = func(terminal) {
+    while let Some((key, success)) = func(&mut execution_context.terminal) {
         if success {
-            return Some((key, success));
+            return Ok(Some((key, success)));
         }
 
-        if !terminal.should_retry() {
-            return Some((key, success));
+        let running = ctrlc::running();
+        if !running {
+            ctrlc::set_running(true);
+        }
+
+        let should_retry = execution_context.terminal.should_retry(running);
+
+        if !ctrlc::running() {
```

Comment:
Maybe `should_retry()` should return `Err` if it was interrupted instead of returning `false`.